### PR TITLE
isTag() additionally check for casing of the first char

### PR DESCRIPTION
**Pull request by asvetliakov**
It's very common to write tests like this:

```jsx
jest.mock("../SomeComponent", () => "MockedComponent");

it("test", () => {
  expect(testRenderer(<Component someParam="a" />)).toMatchSnapshot();
});
```

where ```Component``` may be:
```jsx
import SomeComponent from "./SomeComponent";
export default styled(SomeComponent)` /* stuff */`;
```

In this case ```SomeComponent``` becomes string-typed internally similar to dom-like components and only name with passed props is being rendered is snapshots.

Without this fix, the ```styled-components``` thinks as ```SomeComponent``` is like dom-like component and applies props filtering to it, resulting in filtered out custom props (```someParam``` in example above) from test snapshots.

### CODE REVIEW

File: src/utils/isTag.js [hunk 1]
```diff
@@ -2,5 +2,5 @@
 import type { Target } from '../types';
 
 export default function isTag(target: Target) /* : %checks */ {
-  return typeof target === 'string';
+  return typeof target === 'string' && target.charAt(0) === target.charAt(0).toLowerCase();
```
Comment (probablyup) [inline src/utils/isTag.js]: let's wrap this in a `NODE_ENV !== 'production'` check

Comment (asvetliakov) [reply src/utils/isTag.js]: I think it's not necessary. Built-in tag components are always lower-cased by [react spec](https://reactjs.org/docs/jsx-in-depth.html#user-defined-components-must-be-capitalized)

Comment (probablyup) [reply src/utils/isTag.js]: Yes, I just don't want to include the code in production since it will never happen :)

File: src/utils/generateDisplayName.js [hunk 2]
```diff
@@ -4,5 +4,5 @@ import isTag from './isTag';
 import type { Target } from '../types';
 
 export default function generateDisplayName(target: Target): string {
-  return isTag(target) ? `styled.${target}` : `Styled(${getComponentName(target)})`;
+  return isTag(target) ? `styled.${getComponentName(target)}` : `Styled(${getComponentName(target)})`;
```
Comment (probablyup) [inline src/utils/generateDisplayName.js]: why was this change added? it's not desired

Comment (asvetliakov) [reply src/utils/generateDisplayName.js]: Flow can't coerce correctly to string type for some reason

Comment (probablyup) [reply src/utils/generateDisplayName.js]: Ah ok, just put a `// $FlowFixMe`

Comment (asvetliakov) [reply src/utils/generateDisplayName.js]: done

File: src/utils/getComponentName.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 // @flow
 import type { ComponentType } from 'react';
 
-export default function getComponentName(target: ComponentType<*>): string {
-  return target.displayName || target.name || 'Component';
+export default function getComponentName(target: ComponentType<*> | string): string {
+  return typeof target === "string" ? target : target.displayName || target.name || 'Component';
```
Comment (probablyup) [inline src/utils/getComponentName.js]: this can be reverted as well

**Comment by probablyup** [comment]
Finally, just need a test and a changelog entry :) thanks for your work!

File: src/utils/getComponentName.js (hunk 3 above)
Comment (asvetliakov) [reply src/utils/getComponentName.js]: I don't think so, without it the generated display name will be ```Styled(Component)``` with mocked components because string doesn't have either displayName || name

Comment (probablyup) [reply src/utils/getComponentName.js]: It won't be because you're accounting for the uppercase string name in `isTag`. It'll end up as `styled.YourMockedName`

Comment (asvetliakov) [reply src/utils/getComponentName.js]: No, it won't, because ```isTag``` will return false for such components (check for first lower case letter will fail)

**Comment by asvetliakov** [comment]
@probablyup Ping
